### DISCUSSION TITLE
Logging questions

### DISCUSSION THREAD

**Discussion by shieldz80** [roadrunner-server/roadrunner#2040] [Q&A] [answered]
I am using the Logger plugin to send logs from PHP to the RoadRunner server. Everything works OK, but I do have a couple of questions:

In my RR server configuration the logger plugin is configured like this:

```
logs:
  output: stdout
  encoding: console
  mode: ${RR_LOGGING_MODE:-production}
  level: ${RR_LOGGING_LEVEL:-error}
```
On the PHP side I am using Monolog with a custom handler that sends messages to the RR server

```php
final class RoadRunnerHandler extends AbstractProcessingHandler
{
    public const FORMAT = "%channel%: %message% %context% %extra%\n";

    public function __construct(
        private readonly RoadRunnerLogger $logger,
    ) {
        parent::__construct();
        $this->setFormatter(new LineFormatter(self::FORMAT));
    }

    protected function write(LogRecord $record): void
    {
        $message = $record['formatted'];
        assert(is_string($message) || $message instanceof Stringable);

        $level = $record['level'] instanceof Level ? $record['level']->value : $record['level'];

        match ($level) {
            LogLevel::ERROR, LogLevel::CRITICAL, LogLevel::ALERT, LogLevel::EMERGENCY => $this->logger->error($message),
            LogLevel::WARNING => $this->logger->warning($message),
            LogLevel::INFO, LogLevel::NOTICE => $this->logger->info((string) $message),
            default => $this->logger->debug($message),
        };
    }
}
```
This is pretty much a tweaked implementation of the RoadRunner Handler in the Spiral Framework. The app logs being produced by the RR server look something like this in production:

`1730271338634218848     ERROR   app     main: Some error message here [] []`

In development app logs look something like:

`2024-10-30T07:03:15+0000        DEBUG   app             main: Some debug message here [] []`

So finally, my questions are:

1. Is it possible to have a value like `2024-10-30T07:03:15+0000` instead `1730271338634218848` when the logger plugin is configured with `mode: production`? The unix timestamp is not really helpful when reading logs. For sure I can transform it into what I want, I just want to know if I can avoid that "work".
2. Furthermore, is it possible to have a "timezoned" value? For example: instead of `2024-10-30T07:03:15+0000` to have `2024-10-30T09:03:15+0200`. I suppose this would imply having the possibility to configure the timezone for the RR server or for the RR server to somehow infer it from environment.
3. Finally, would it be possible to have logs like `2024-10-30T07:03:15+0000        DEBUG   app.main: Some debug message here [] []`? Not a big deal of course, just nice to have. I suppose this would imply having the possibility of controlling the logger plugin's logging format. 

Thanks in advance.

**Comment by rustatian** [comment, marked as answer]
Hey @shieldz80 👋 
1. No, this is not configurable in the production mode, but you may fork the logger plugin, use any encoder for the timestamps you need and build your custom RR instance with your modified logger.
2. All logs are UTC. Also, possible if you fork the logger plugin.
3. In development mode, this is not a space ` app             main`, but a color code. Because logs in the development mode are colored. So, as in previous answers, everything is possible if you fork logger plugin and tune it for yourself: https://docs.roadrunner.dev/docs/customization/build

**Reply by shieldz80** [reply]
Thanks @rustatian
